build/bazel/remote/fetch/v1/remote_fetch.proto

```diff
@@ -0,0 +1,442 @@
+// uuid namespace identifier: urn:uuid:f81d4fae-7dec-11d0-a765-00a0c91e6bf6.
+// This is most applicable to named content that is Push’d, where the URN
+// serves as an agreed-upon key, but carries no other inherent meaning.
+//
+// Service implementations may choose to support only URLs, only URNs for
+// Push’d content, only other URIs for which the server and client agree upon
+// semantics of, or any mixture of the above.
+
+// Qualifiers are used to disambiguate or sub-select content that shares a URI.
+// This may include specifying a particular commit or branch, in the case of
+// URIs referencing a repository; they could also be used to specify a
+// particular subdirectory of a repository or tarball. Qualifiers may also be
+// used to ensure content matches what the client expects, even when there is
+// no ambiguity to be had - for example, a qualifier specifying a checksum
+// value.
+//
+// In cases where the semantics of the request are not immediately clear from
+// the URL and/or qualifiers - e.g. dictated by URL scheme - it is recommended
+// to use an additional qualifier to remove the ambiguity. The `resource_type`
+// qualifier is recommended for this purpose.
+//
+// Qualifiers may be supplied in any order.
+message Qualifier {
+  // The “name” of the qualifier, for example “resource_type”.
+  // No separation is made between ‘standard’ and ‘nonstandard’
+  // qualifiers, in accordance with https://tools.ietf.org/html/rfc6648,
+  // however implementers *SHOULD* take care to avoid ambiguity.
+  string name = 1;
+
+  // The “value” of the qualifier. Semantics will be dictated by the name.
+  string value = 2;
+}
+
+// The Fetcher service resolves or fetches assets referenced by URI and
+// Qualifiers, returning a Digest for the content in 
+// [ContentAddressableStorage][build.bazel.remote.execution.v2.ContentAddressableStorage].
+//
+// As with other services in the Remote Execution API, any call may return an
+// error with a [RetryInfo][google.rpc.RetryInfo] error detail providing
+// information about when the client should retry the request; clients SHOULD
+// respect the information provided.
+service Fetcher {
+  // Rsolve or fetch referenced assets, making them available to the caller and
+  // other consumers in the [ContentAddressableStorage][build.bazel.remote.execution.v2.ContentAddressableStorage].
+  //
+  // Servers *MAY* fetch content that they do not already have cached, for any
+  // URLs they support.
+  //
+  // Servers *SHOULD* ensure that referenced files are present in the CAS at the
+  // time of the response, and (if supported) that they will remain available
+  // for a reasonable period of time. In the event that a client receives a
+  // reference to content that is no longer present, it *MAY* re-issue the
+  // request with `oldest_content_accepted` set to a more recent timestamp than
+  // the original attempt, to induce a re-fetch from origin.
```

Comment:
@jmillikin-stripe, I believe this is one of the few places where the proposals differ, as per https://github.com/bazelbuild/proposals/pull/160/files#diff-3bdfd5c462602da5c73b9b00c1696223R116

Is there a reason you made the guarantee stronger?

Reply:
The fetch API isn't very useful if the server is allowed to reply "OK I have added your data to the CAS" when it has, in fact, _not_ added that data. There's no way for the client to handle that situation gracefully, and in Bazel it causes a big stack trace if downloading a just-enCASed file fails.

Reply:
The text is actually inspired by [ActionCache.GetActionResult](https://github.com/bazelbuild/remote-apis/blob/master/build/bazel/remote/execution/v2/remote_execution.proto#L140).  I would like for both to be **MUST**, but realistically, this is going to be hard to guarantee without amending the [ContentAddressableStorage](https://github.com/bazelbuild/remote-apis/blob/master/build/bazel/remote/execution/v2/remote_execution.proto#L234) lifetime requirements.  As such we can only suggest a strong **SHOULD**.

Clients should always handle errors gracefully, at least more graceful than a big stack trace like a helpful error message to the user.  The client can also retry, or fall back to direct download if allowed.

Reply:
I've updated the language slightly, to include that the server should extend lifetime of referenced blobs where necessary and applicable, mirroring the language of ActionCache.GetActionResult more closely.